### PR TITLE
[AIRFLOW-3702] Add backfill option to run backwards

### PR DESCRIPTION
**Pull request by dima-asana**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following:
  - https://issues.apache.org/jira/browse/AIRFLOW-3702

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This adds an optional capability for the backfill CLI command to process dates in reverse order.

### Tests

- [x] My PR adds the following unit tests:
test_jobs::BackfillJobTest::test_backfill_enqueue_backwards

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by feng-tao** [comment]
@dima-asana , I have been pretty busy lately. Looking at your code,  I think your pr is almost there which only needs to handle the case when the DAG has depend_on_past(we could throw the exception if reverse_backfill and depend_on_past are both true).

Do you want to update your pr ? I think we could commit yours and close mine. What do you think?

**Comment by dima-asana** [comment]
Sure, updated.  Note that in addition to the suggested change for depends_on_past handling, I changed the executor queue dict to an ordered dict.  I think for this change to be impactful that's necessary -- I didn't realize that the executor queue were orderless until testing this code more.

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -1782,6 +1783,13 @@ class CLIFactory(object):
                 "all the failed tasks for the backfill date range "
                 "instead of throwing exceptions"),
             "store_true"),
+        'run_backwards': Arg(
+            ("-B", "--run_backwards",),
+            (
+                "if set, the backfill will run tasks from the most "
+                "recent day first "
+                "instead of throwing exceptions"),
```
Comment (feng-tao) [inline airflow/bin/cli.py]: How about adding one more line "this option will fail if the DAG depends on past."?

**Review by feng-tao — APPROVED** [review]
1 small nit. LGTM. thanks for the patch.

File: airflow/bin/cli.py (hunk 1 above)
Comment (dima-asana) [reply airflow/bin/cli.py]: done

**Comment by damnko** [comment]
Hi, can the `run_backwards` parameter be used when defining a DAG like so?
```python
from airflow import DAG
dag = DAG(
    dag_id = "..."
    scheduled_interval = "@daily"
    run_backwards = True
)
```

It seems to be something that can be used only manually through the Airflow cli but wanted to confirm. Is there any way to setup a DAG to run as backfilling job from the most recent DAG to the oldest?
Thanks

**Comment by aashayVimeo** [comment]
hi @dima-asana @feng-tao Any updates on the above comment for running backwards dag from dag definition.

```
from airflow import DAG
dag = DAG(
    dag_id = "...",
    scheduled_interval = "@daily",
    run_backwards = True,
    ...
)
```

**Pull request merged**